taichi/transforms/constant_fold.cpp

```diff
@@ -176,30 +149,76 @@ class ConstantFold : public BasicStmtVisitor {
     auto operand = stmt->operand->cast<ConstStmt>();
     if (!operand)
       return;
-    if (stmt->is_cast()) {
-      bool cast_available = true;
+    if (stmt->is_cast() && stmt->op_type == UnaryOpType::cast_bits) {
       TypedConstant new_constant(stmt->ret_type);
-      if (stmt->op_type == UnaryOpType::cast_bits) {
-        new_constant.value_bits = operand->val.value_bits;
-      } else {
-        if (stmt->cast_type == PrimitiveType::f32) {
-          new_constant.val_f32 = float32(operand->val.val_cast_to_float64());
-        } else if (stmt->cast_type == PrimitiveType::f64) {
-          new_constant.val_f64 = operand->val.val_cast_to_float64();
-        } else {
-          cast_available = false;
-        }
-      }
-      if (cast_available) {
-        insert_and_erase(stmt, new_constant);
-        return;
-      }
-    }
-    auto dst_type = stmt->ret_type;
-    TypedConstant new_constant(dst_type);
-    if (jit_evaluate_unary_op(new_constant, stmt, operand->val)) {
+      new_constant.value_bits = operand->val.value_bits;
       insert_and_erase(stmt, new_constant);
+      return;
     }
+    const auto dt = operand->val.dt;
+    if (!is_good_type(dt))
+      return;
+    const auto dst_type = stmt->ret_type;
+    switch (stmt->op_type) {
+#define HANDLE_REAL_AND_INTEGRAL_UNARY(OP_TYPE, OP_CPP)                     \
+  case UnaryOpType::OP_TYPE: {                                              \
+    if (dt->is_primitive(PrimitiveTypeID::f32) ||                           \
```

Comment:
We can use `is_real`, `is_signed`, and `is_unsigned` functions here

Reply:
Yup the reason that we list types explicitly here is to align with `is_good_type` but we can update these for sure!